plugin/storage/es/spanstore/reader_test.go

```diff
@@ -484,38 +454,367 @@ func TestSpanReader_bucketToStringArrayError(t *testing.T) {
-	return searchService
+	return searchService.On("Do", mock.AnythingOfType("*context.emptyCtx"))
+}
+
+func TestTraceQueryParameterValidation(t *testing.T) {
+	var malformedtsp *spanstore.TraceQueryParameters
+	err := validateQuery(malformedtsp)
+	assert.EqualError(t, err, ErrMalformedRequestObject.Error())
+
+	tsp := &spanstore.TraceQueryParameters{
+		ServiceName: "",
+		Tags: map[string]string{
+			"michael": "jackson",
+		},
+	}
+	err = validateQuery(tsp)
+	assert.EqualError(t, err, ErrServiceNameNotSet.Error())
+
+	tsp.ServiceName = serviceName
+
+	tsp.StartTimeMin = time.Time{} //time.Unix(0,0) doesn't work because timezones
+	tsp.StartTimeMax = time.Time{}
+	err = validateQuery(tsp)
+	assert.EqualError(t, err, ErrStartAndEndTimeNotSet.Error())
+
+	tsp.StartTimeMin = time.Now()
+	tsp.StartTimeMax = time.Now().Add(-1 * time.Hour)
+	err = validateQuery(tsp)
+	assert.EqualError(t, err, ErrStartTimeMinGreaterThanMax.Error())
+
+	tsp.StartTimeMin = time.Now().Add(-1 * time.Hour)
+	tsp.StartTimeMax = time.Now()
+	err = validateQuery(tsp)
+	assert.Nil(t, err)
+
+	tsp.DurationMin = time.Hour
+	tsp.DurationMax = time.Minute
+	err = validateQuery(tsp)
+	assert.EqualError(t, err, ErrDurationMinGreaterThanMax.Error())
+}
+
+func TestSpanReader_buildTraceIDAggregation(t *testing.T) {
+	expectedStr := `{ "terms": { "size": 123, "field": "` + traceIDField + `" }}`
+	withSpanReader(func(r *spanReaderTest) {
+		traceIDAggregation := r.reader.buildTraceIDAggregation(123)
+		actual, err := traceIDAggregation.Source()
+		require.NoError(t, err)
+
+		expected := make(map[string]interface{})
+		json.Unmarshal([]byte(expectedStr), &expected)
+		expected["terms"].(map[string]interface{})["size"] = 123
+		assert.EqualValues(t, expected, actual)
+	})
+}
+
+func TestSpanReader_buildFindTraceIDsQuery(t *testing.T) {
+	withSpanReader(func(r *spanReaderTest) {
+		traceQuery := &spanstore.TraceQueryParameters{
+			DurationMin:   time.Second,
+			DurationMax:   time.Second * 2,
+			StartTimeMin:  time.Time{},
+			StartTimeMax:  time.Time{}.Add(time.Second),
+			ServiceName:   "s",
+			OperationName: "o",
+			Tags: map[string]string{
+				"michael": "jackson",
+			},
+		}
+
+		actualQuery := r.reader.buildFindTraceIDsQuery(traceQuery)
+		actual, err := actualQuery.Source()
+		require.NoError(t, err)
+		expectedQuery := elastic.NewBoolQuery().
+			Must(
+				r.reader.buildDurationQuery(time.Second, time.Second*2),
+				r.reader.buildStartTimeQuery(time.Time{}, time.Time{}.Add(time.Second)),
+				r.reader.buildServiceNameQuery("s"),
+				r.reader.buildOperationNameQuery("o"),
+				r.reader.buildTagQuery("michael", "jackson"),
+			)
+		expected, err := expectedQuery.Source()
+		require.NoError(t, err)
+		assert.Equal(t, expected, actual)
+	})
+}
+
+func TestSpanReader_buildDurationQuery(t *testing.T) {
+	// { "range":  { "duration": { "gte": 1000000, "lte": 2000000 }}}
```

Comment:
rm? curious though, "gte" gets converted to "from" by the driver? Is this accepted by ES?